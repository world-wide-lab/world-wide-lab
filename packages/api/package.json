{
  "name": "@world-wide-lab/api",
  "version": "0.0.1",
  "description": "",
  "main": "dist/index.js",
  "scripts": {
    "test": "jest"
  },
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@babel/preset-env": "^7.22.9",
    "@babel/preset-typescript": "^7.22.5",
    "@types/jest": "^29.5.2",
    "@types/node": "^20.3.1",
    "babel-jest": "^29.5.0",
    "jest": "^29.5.0",
    "typescript": "^5.1.3"
  }
}
