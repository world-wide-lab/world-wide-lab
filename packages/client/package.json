{
  "name": "@world-wide-lab/client",
  "version": "0.4.0",
  "description": "",
  "repository": {
    "type": "git",
    "url": "https://github.com/world-wide-lab/world-wide-lab.git",
    "directory": "packages/client"
  },
  "main": "dist/index.js",
  "scripts": {
    "prepack": "npm run build",
    "prebuild": "node -p \"'export const VERSION = ' + JSON.stringify(require('./package.json').version) + ';'\" > src/version.ts",
    "build": "tsc -p tsconfig-build.json",
    "test": "vitest run",
    "test:build": "CLIENT=build npm run build && npm run test"
  },
  "author": "",
  "license": "MIT",
  "devDependencies": {
    "@babel/preset-env": "^7.22.9",
    "@babel/preset-typescript": "^7.22.5",
    "@types/node": "^20.3.1",
    "vitest": "^1.6.0",
    "typescript": "^5.1.3"
  }
}
