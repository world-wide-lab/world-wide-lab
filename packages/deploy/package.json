{
  "name": "@world-wide-lab/deploy",
  "license": "MIT",
  "description": "Deploy World-Wide-Lab to the cloud.",
  "version": "0.2.0",
  "main": "dist/index.js",
  "scripts": {
    "prepack": "npm run build",
    "build": "tsc"
  },
  "devDependencies": {
    "@types/node": "^22.9.0"
  },
  "dependencies": {
    "@pulumi/aws": "^6.57.0",
    "@pulumi/awsx": "^2.16.1",
    "@pulumi/azure-native": "^2.83.0",
    "@pulumi/pulumi": "^3.137.0",
    "deepmerge": "^4.3.1",
    "dotenv": "^16.4.5",
    "typescript": "^5.6.3"
  }
}
