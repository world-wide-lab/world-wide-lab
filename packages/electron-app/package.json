{
  "name": "@world-wide-lab/electron-app",
  "version": "0.0.1",
  "description": "",
  "main": "index.js",
  "scripts": {
    "dev": "electron .",
    "build:electron": "npm run build:wwl && npm run package",
    "build:dev": "npm run build:wwl && npm run package:dev",
    "build:wwl": "npm run build --prefix ../server/",
    "package": "electron-builder",
    "package:dev": "electron-builder --dir",
    "test:playwright-wip": "../../node_modules/@playwright/test/cli.js test"
  },
  "author": "",
  "license": "NO_LICENSE_YET",
  "dependencies": {
    "@world-wide-lab/server": "file:../server"
  },
  "devDependencies": {
    "@playwright/test": "^1.36.2",
    "electron": "^24.6.1",
    "electron-builder": "^24.4.0",
    "playwright": "^1.36.2"
  }
}
