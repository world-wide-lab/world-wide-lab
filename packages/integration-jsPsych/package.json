{
  "name": "@world-wide-lab/integration-jspsych",
  "version": "0.7.3",
  "description": "",
  "type": "module",
  "main": "dist/index.cjs",
  "exports": {
    "import": "./dist/index.js",
    "require": "./dist/index.cjs"
  },
  "typings": "dist/index.d.ts",
  "unpkg": "dist/index.browser.min.js",
  "files": ["src", "dist"],
  "source": "src/index.ts",
  "scripts": {
    "prepack": "npm run build",
    "prebuild": "node -p \"'export const VERSION = ' + JSON.stringify(require('./package.json').version) + ';'\" > src/version.ts",
    "test": "jest",
    "test:watch": "npm test -- --watch",
    "tsc": "tsc",
    "build": "rollup --config",
    "build:watch": "npm run build -- --watch"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/world-wide-lab/world-wide-lab.git",
    "directory": "packages/integration-jsPsych"
  },
  "author": "Jan Simson",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/world-wide-lab/world-wide-lab/issues"
  },
  "homepage": "https://github.com/world-wide-lab/world-wide-lab/tree/main/packages/extension-template-ts",
  "peerDependencies": {
    "jspsych": ">=7.0.0"
  },
  "devDependencies": {
    "@jspsych/config": "^2.0.2",
    "@jspsych/plugin-html-keyboard-response": "^2.0.0",
    "@jspsych/test-utils": "1.1.2",
    "@types/jest": "^29.5.12",
    "@world-wide-lab/server": "file:../server",
    "jspsych": "^7.0.0",
    "ts-jest": "^29.1.1"
  },
  "dependencies": {
    "@world-wide-lab/client": "^0.7.1"
  }
}
