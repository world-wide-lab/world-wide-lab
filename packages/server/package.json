{
  "name": "@world-wide-lab/server",
  "version": "0.2.3",
  "description": "",
  "main": "src/index.ts",
  "scripts": {
    "dev": "ts-node-dev --respawn --pretty --transpile-only src/index.ts",
    "build": "tsc -p tsconfig-build.json",
    "start": "node dist/index.js",
    "start:build": "npm run build && npm start",
    "test": "jest",
    "start:playwright": "WWL_ENV_FILE=.env.playwright npm run build && node dist/index.js",
    "test:playwright": "WWL_ENV_FILE=.env.playwright npx playwright test",
    "update-docs-screenshots": "WWL_ENV_FILE=.env.playwright ALLOW_NON_TESTS=1 npx playwright test update-docs-screenshots"
  },
  "keywords": [],
  "author": "",
  "license": "NO_LICENSE_YET",
  "dependencies": {
    "@adminjs/express": "^5.1.0",
    "@adminjs/sequelize": "^3.0.0",
    "@types/recharts": "^1.8.24",
    "adminjs": "^6.8.7",
    "connect-session-sequelize": "^7.1.7",
    "cors": "^2.8.5",
    "dotenv": "^16.1.4",
    "express": "^4.18.2",
    "express-formidable": "^1.2.0",
    "express-session": "^1.17.3",
    "frappe-charts": "^1.6.2",
    "helmet": "^7.0.0",
    "highlight.js": "^11.8.0",
    "json-2-csv": "^5.0.1",
    "pg": "^8.11.0",
    "pg-copy-streams": "^6.0.6",
    "react-frappe-charts": "^4.1.0",
    "sequelize": "^6.32.1",
    "sqlite3": "^5.1.6",
    "swagger-jsdoc": "^6.2.8",
    "swagger-ui-express": "^4.6.3",
    "tslib": "^2.6.0",
    "umzug": "^3.2.1",
    "validator": "^13.9.0",
    "winston": "^3.9.0",
    "yup": "^1.3.3"
  },
  "devDependencies": {
    "@babel/core": "^7.22.5",
    "@babel/preset-env": "^7.22.5",
    "@babel/preset-typescript": "^7.22.5",
    "@playwright/test": "^1.36.2",
    "@types/cors": "^2.8.13",
    "@types/dotenv": "^8.2.0",
    "@types/express": "^4.17.17",
    "@types/express-session": "^1.17.7",
    "@types/helmet": "^4.0.0",
    "@types/jest": "^29.5.2",
    "@types/node": "^20.3.1",
    "@types/pg": "^8.10.8",
    "@types/pg-copy-streams": "^1.2.5",
    "@types/styled-components": "^5.1.26",
    "@types/supertest": "^2.0.12",
    "@types/swagger-jsdoc": "^6.0.1",
    "@types/swagger-ui-express": "^4.1.3",
    "babel-jest": "^29.5.0",
    "jest": "^29.5.0",
    "nodemon": "^2.0.22",
    "supertest": "^6.3.3",
    "ts-node-dev": "^2.0.0",
    "typescript": "^5.1.3"
  }
}
