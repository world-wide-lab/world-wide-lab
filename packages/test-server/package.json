{
  "name": "@world-wide-lab/test-server",
  "version": "0.1.0",
  "description": "Package for running tests against the World-Wide-Lab server",
  "scripts": {
    "test": "vitest run"
  },
  "keywords": [],
  "author": "",
  "type": "module",
  "license": "MIT",
  "devDependencies": {
    "@babel/preset-env": "^7.22.9",
    "@babel/preset-typescript": "^7.22.5",
    "@types/node": "^20.3.1",
    "@types/supertest": "^2.0.12",
    "@world-wide-lab/server": "file:../server",
    "babel-jest": "^29.5.0",
    "supertest": "^6.3.3",
    "typescript": "^5.1.3",
    "vitest": "^1.6.0"
  }
}
