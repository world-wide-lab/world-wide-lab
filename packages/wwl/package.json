{
  "name": "@world-wide-lab/wwl",
  "version": "0.0.1",
  "description": "",
  "main": "src/index.ts",
  "scripts": {
    "dev": "ts-node-dev --respawn --pretty --transpile-only src/index.ts",
    "build": "tsc -p tsconfig-build.json",
    "start": "node dist/index.js",
    "test": "jest"
  },
  "keywords": [],
  "author": "",
  "license": "NO_LICENSE_YET",
  "dependencies": {
    "@adminjs/express": "^5.1.0",
    "@adminjs/sequelize": "^3.0.0",
    "adminjs": "^6.8.7",
    "connect-session-sequelize": "^7.1.7",
    "cors": "^2.8.5",
    "dotenv": "^16.1.4",
    "express": "^4.18.2",
    "express-formidable": "^1.2.0",
    "express-session": "^1.17.3",
    "helmet": "^7.0.0",
    "pg": "^8.11.0",
    "sequelize": "^6.32.1",
    "sqlite3": "^5.1.6",
    "swagger-jsdoc": "^6.2.8",
    "swagger-ui-express": "^4.6.3",
    "tslib": "^2.6.0",
    "umzug": "^3.2.1",
    "winston": "^3.9.0"
  },
  "devDependencies": {
    "@babel/core": "^7.22.5",
    "@babel/preset-env": "^7.22.5",
    "@babel/preset-typescript": "^7.22.5",
    "@types/cors": "^2.8.13",
    "@types/dotenv": "^8.2.0",
    "@types/express": "^4.17.17",
    "@types/express-session": "^1.17.7",
    "@types/helmet": "^4.0.0",
    "@types/jest": "^29.5.2",
    "@types/node": "^20.3.1",
    "@types/styled-components": "^5.1.26",
    "@types/supertest": "^2.0.12",
    "@types/swagger-jsdoc": "^6.0.1",
    "@types/swagger-ui-express": "^4.1.3",
    "babel-jest": "^29.5.0",
    "jest": "^29.5.0",
    "nodemon": "^2.0.22",
    "supertest": "^6.3.3",
    "ts-node-dev": "^2.0.0",
    "typescript": "^5.1.3"
  }
}
