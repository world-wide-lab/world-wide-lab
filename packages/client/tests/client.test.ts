import "./setup_env";

import { version as packageVersion } from "../package.json";

import {
  Client as DevClient,
  Participant,
  Session,
  WorldWideLabError,
} from "../src";

import {
  type Server,
  init as initDev,
} from "@world-wide-lab/server/src/init.ts";

const Client = process.env.CLIENT === "build" ? import("../dist") : DevClient;

const API_KEY = process.env.DEFAULT_API_KEY;

describe("Client", () => {
  let server: Server;
  let client: DevClient;
  beforeAll(async () => {
    server = await initDev();

    // @ts-ignore - We know that the server will only be returned after listen() is finished
    client = new Client({ url: `http://localhost:${server.address().port}` });

    global.fetch = vi.fn(fetch);
  }, 10000);
  afterAll(async () => {
    await new Promise((resolve, reject) => {
      server.close((error) => {
        if (error) {
          reject(error);
        } else {
          resolve(undefined);
        }
      });
    });
  }, 10000);

  it("should create a new participant", async () => {
    const participant = await client.createParticipant();

    expect(participant instanceof Participant).toBe(true);
    expect(participant.participantId).toBeDefined();
  });

  it("should start a new session (without a linked participant)", async () => {
    global.window = {
      location: {
        href: "https://worldwidelab.org/?myId=123&otherId=test",
        search: "?myId=123&otherId=test",
      },
    } as any;

    const session = await client.createSession({ studyId: "example" });

    expect(session instanceof Session).toBe(true);
    expect(session.sessionId).toBeDefined();

    // @ts-ignore Complicated to get typescript to acknowledge that fetch is indeed also a mock function now
    const sessionParams = JSON.parse(global.fetch.mock.calls.pop()[1].body);

    // Version should be up to date
    expect(sessionParams.clientMetadata.version).toBe(packageVersion);

    sessionParams.clientMetadata.version = "VERSION";
    expect(sessionParams).toMatchSnapshot();
  });

  it("should start a new session (with a linked participant)", async () => {
    const participant = await client.createParticipant();
    const session = await client.createSession({
      studyId: "example",
      participant,
    });

    expect(session instanceof Session).toBe(true);
    expect(session.sessionId).toBeDefined();
    expect(session.participant instanceof Participant).toBe(true);
    expect(session.participant?.participantId).toBeDefined();
  });

  it("should throw an error when session creation fails", async () => {
    await expect(async () => {
      const session = await client.createSession({
        studyId: "non-existent-study",
      });
    }).rejects.toThrow("Failed to initialize Session");
  });

  it("should store responses", async () => {
    const studyId = "studyId-check-response-storage";

    // Create a new study
    const createStudyResponseJson = await (
      await client.call("POST", "/study/", { studyId })
    ).json();
    expect(createStudyResponseJson.studyId).toBe(studyId);
    expect(Object.keys(createStudyResponseJson)).toMatchSnapshot();

    const session = await client.createSession({ studyId });

    expect(
      await session.response({
        name: "example_name",
        payload: { ex_key: "ex_value" },
      }),
    ).toBe(true);
    expect(
      await session.response({
        name: undefined,
        payload: { key_n1: "no-name", numerical: 1 },
      }),
    ).toBe(true);

    // Download export from database
    const responseJson = await (
      await client.call(
        "GET",
        `/study/${studyId}/data/responses-raw/json`,
        undefined,
        {
          // Manually set headers to add authorization one
          headers: {
            ContentType: "application/json",
            Authorization: `Bearer ${API_KEY}`,
          },
        },
      )
    ).json();

    expect(responseJson.length).toBe(2);

    const variyngKeys = ["responseId", "sessionId", "updatedAt", "createdAt"];

    for (const response of responseJson) {
      for (const key of variyngKeys) {
        response[key] = "overwritten-as-it-varies";
      }
    }
    expect(responseJson).toMatchSnapshot();
    expect(Object.keys(responseJson)).toMatchSnapshot();
  });

  it("should store and retrieve participant data", async () => {
    const participant = await client.createParticipant();

    const participantUpdateResult = await participant.update({
      privateInfo: {
        name: "John Doe",
      },
      publicInfo: {
        condition: "A",
      },
    });
    expect(participantUpdateResult).toBe(true);

    const publicParticipantInfo = await participant.getPublicInfo();
    expect(publicParticipantInfo.publicInfo.condition).toBe("A");
  });

  it("should store participant data upon creation", async () => {
    const participant = await client.createParticipant({
      privateInfo: {
        name: "John Doe",
      },
      publicInfo: {
        condition: "B",
      },
    });

    const publicParticipantInfo = await participant.getPublicInfo();
    expect(publicParticipantInfo.publicInfo.condition).toBe("B");
  });

  it("should store and retrieve session data", async () => {
    const session = await client.createSession({ studyId: "example" });

    const sessionUpdateResult = await session.update({
      privateInfo: {
        name: "Session No Uno",
      },
      publicInfo: {
        condition: "A",
      },
    });
    expect(sessionUpdateResult).toBe(true);

    const publicSessionInfo = await session.getPublicInfo();
    expect(publicSessionInfo.publicInfo.condition).toBe("A");
  });

  it("should store session data upon creation", async () => {
    const session = await client.createSession({
      studyId: "example",

      privateInfo: {
        name: "Session No Duo",
      },
      publicInfo: {
        condition: "C",
      },
    });

    const publicSessionInfo = await session.getPublicInfo();
    expect(publicSessionInfo.publicInfo.condition).toBe("C");
  });

  it("should finish a session", async () => {
    const session = await client.createSession({ studyId: "example" });

    const sessionFinishResult = await session.finish();
    expect(sessionFinishResult).toBe(true);
  });

  it("should add & retrieve scores from the leaderboard", async () => {
    const session = await client.createSession({ studyId: "example" });

    const tsBeforeScores = new Date();

    const addScoreResult = await session.addScoreToLeaderboard("lb-test", {
      score: 1337,
      publicIndividualName: "Kevin Flynn",
      publicGroupName: "Encom",
    });
    expect(addScoreResult).toBeDefined();
    const addScoreResult2 = await session.addScoreToLeaderboard("lb-test", {
      score: 663,
      publicIndividualName: "Sam Flynn",
      publicGroupName: "Encom",
    });
    expect(addScoreResult2).toBeDefined();

    const getIndividualScoresResult =
      await client.getLeaderboardScores("lb-test");
    expect(getIndividualScoresResult).toMatchSnapshot();

    const getGroupScoresResult = await client.getLeaderboardScores(
      "lb-test",
      "groups",
      { aggregate: "sum" },
    );
    expect(getGroupScoresResult).toMatchSnapshot();

    const tsFuture = new Date();
    tsFuture.setSeconds(tsFuture.getSeconds() + 10);

    // Older timestamp should work just fine
    const getIndividualScoresResultWithTimestamp =
      await client.getLeaderboardScores("lb-test", "individual", {
        updatedAfter: tsBeforeScores,
      });
    expect(getIndividualScoresResultWithTimestamp).toMatchObject(
      getIndividualScoresResult,
    );

    // Future timestamp should return empty leaderboard
    const getIndividualScoresResultInFuture = await client.getLeaderboardScores(
      "lb-test",
      "individual",
      { updatedAfter: tsFuture },
    );
    expect(getIndividualScoresResultInFuture).toMatchObject({ scores: [] });
  });

  it("should update a score on the leaderboard", async () => {
    const session = await client.createSession({ studyId: "example" });

    const leaderboardScoreId = await session.addScoreToLeaderboard("lb-test", {
      score: 1337,
      publicIndividualName: "Legolas",
    });
    expect(leaderboardScoreId).toBeDefined();

    const updateScoreResult = await session.updateLeaderboardScore(
      "lb-test",
      leaderboardScoreId,
      {
        score: 1500,
        publicIndividualName: "Gimli",
        publicGroupName: "Die Gefährten",
      },
    );
    expect(updateScoreResult).toBe(true);

    const getIndividualScoresResult =
      await client.getLeaderboardScores("lb-test");

    expect(getIndividualScoresResult.scores).toEqual(expect.arrayContaining([
      {
        score: 1500,
        publicIndividualName: "Gimli",
        publicGroupName: "Die Gefährten",
      },
    ]));
    expect([
      {
        score: 1337,
        publicIndividualName: "Legolas",
      },
    ]).not.toEqual(expect.arrayContaining(getIndividualScoresResult.scores));
    expect([
      {
        score: 1337,
        publicIndividualName: "Legolas",
        publicGroupName: "Die Gefährten",
      },
    ]).not.toEqual(expect.arrayContaining(getIndividualScoresResult.scores));
  });

  it("should return the leaderboardScoreId when adding a score to the leaderboard", async () => {
    const session = await client.createSession({ studyId: "example" });

    const leaderboardScoreId = await session.addScoreToLeaderboard("lb-test", {
      score: 1337,
      publicIndividualName: "Kevin Flynn",
      publicGroupName: "Encom",
    });
    expect(leaderboardScoreId).toBeDefined();
    expect(typeof leaderboardScoreId).toBe("number");
  });
});
