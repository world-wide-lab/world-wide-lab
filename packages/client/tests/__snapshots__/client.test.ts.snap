// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Client > should start a new session (without a linked participant) 1`] = `
{
  "clientMetadata": {
    "library": "@world-wide-lab/client",
    "searchParams": {
      "myId": "123",
      "otherId": "test",
    },
    "url": "https://worldwidelab.org/?myId=123&otherId=test",
    "version": "VERSION",
  },
  "studyId": "example",
}
`;

exports[`Client > should store responses 1`] = `
[
  "success",
  "studyId",
]
`;

exports[`Client > should store responses 2`] = `
[
  {
    "Session.participantId": null,
    "createdAt": "overwritten-as-it-varies",
    "name": "example_name",
    "payload": "{"ex_key":"ex_value"}",
    "responseId": "overwritten-as-it-varies",
    "sessionId": "overwritten-as-it-varies",
    "updatedAt": "overwritten-as-it-varies",
  },
  {
    "Session.participantId": null,
    "createdAt": "overwritten-as-it-varies",
    "name": null,
    "payload": "{"key_n1":"no-name","numerical":1}",
    "responseId": "overwritten-as-it-varies",
    "sessionId": "overwritten-as-it-varies",
    "updatedAt": "overwritten-as-it-varies",
  },
]
`;

exports[`Client > should store responses 3`] = `
[
  "0",
  "1",
]
`;
